cpio/cpio.c

```diff
@@ -782,13 +782,52 @@ file_to_archive(struct cpio *cpio, const char *srcpath)
 	return (r);
 }
 
+/*
+ * This functions tries to do an in-kernel copy of a file.  If it fails, it
+ * falls-back to a copy loop.
+ */
+static void
+copy_file(struct cpio *cpio, int fd_in)
+{
+	ssize_t bytes_read, bytes_write;
+#ifdef HAVE_COPY_FILE_RANGE
+	struct stat s;
+
+	if (fstat(fd_in, &s) != 0)
+		lafe_errc(1, errno, "Can't stat file");
+
+	bytes_write = archive_write_disk_copy_file(cpio->archive, fd_in,
+						   s.st_size);
+	if (bytes_write == s.st_size)
+		return;
+	if (bytes_write == ARCHIVE_FAILED)
+		lafe_errc(1, archive_errno(cpio->archive), "%s",
+			  archive_error_string(cpio->archive));
+#endif
+	/* Fallback to a copy loop in case of partial copy or if copy failed */
```

Comment:
This next block of code should be part of `archive_write_disk_copy_file`.

Reply:
Yeah, to be honest that was my initial intention.  I tried to do that initially but found out that things would be a bit more complicated.  For example, the loop uses the cpio->buff and ->buff_size; moving it into `archive_write_disk_copy_file` would mean to re-implement that copy logic.  Obviously, not a show-stopper.  I'll rethink it and try to come up with something different.